TF 2.0.0-alpha0 doesn't build on FreeBSD: error in hwloc

**System information**
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): FreeBSD 12.0
- Bazel version (if compiling from source): 0.23.0
- GCC/Compiler version (if compiling from source): Clang 8.0.0

I'n trying to build TensorFlow from source on FreeBSD 12. During build the following problem shows up:

```
ERROR: /usr/home/arr/.cache/bazel/_bazel_arr/5d34786d346adb404a81c63780107a71/external/hwloc/BUILD.bazel:212:1: C++ compilation of rule '@hwloc//:hwloc' failed (Exit 1)
external/hwloc/hwloc/topology-linux.c:39:10: fatal error: 'mntent.h' file not found
#include <mntent.h>
         ^~~~~~~~~~
1 error generated.
```

This is because `topology-linux.c` is being compiled instead of `topology-freebsd.c`. I have zero knowledge in Bazel, so I wasn't able to include on of these files based on the condition.

This should be easy to fix for anyone experienced with Bazel.

--- a/tensorflow/BUILD
+++ b/tensorflow/BUILD
@@ -497,6 +497,10 @@ tf_cc_shared_object(
     linkopts = select({
         "//tensorflow:macos": [],
         "//tensorflow:windows": [],
+        "//tensorflow:freebsd": [
+            "-Wl,--version-script,$(location //tensorflow:tf_framework_version_script.lds)",
+            "-lexecinfo"
+        ],
         "//conditions:default": [
             "-Wl,--version-script,$(location //tensorflow:tf_framework_version_script.lds)",
         ],

--- a/tensorflow/compiler/xla/BUILD
+++ b/tensorflow/compiler/xla/BUILD
@@ -165,6 +165,10 @@ cc_library(
         "statusor.h",
     ],
     visibility = ["//visibility:public"],
+    linkopts = select({
+        "@org_tensorflow//tensorflow:freebsd": ["-lexecinfo"],
+        "//conditions:default": []
+    }),
     deps = [
         ":status",
         "//tensorflow/stream_executor/lib",

--- a/tensorflow/contrib/ignite/kernels/client/ignite_plain_client_unix.cc
+++ b/tensorflow/contrib/ignite/kernels/client/ignite_plain_client_unix.cc
@@ -16,6 +16,7 @@ limitations under the License.
 #include "tensorflow/contrib/ignite/kernels/client/ignite_plain_client.h"
 
 #include <arpa/inet.h>
+#include <netinet/in.h>
 #include <netdb.h>
 #include <sys/socket.h>
 #include <unistd.h>

--- a/tensorflow/core/platform/cloud/gcs_dns_cache.cc
+++ b/tensorflow/core/platform/cloud/gcs_dns_cache.cc
@@ -16,6 +16,8 @@ limitations under the License.
 #include "tensorflow/core/platform/cloud/gcs_dns_cache.h"
 #ifndef _WIN32
 #include <arpa/inet.h>
+#include <netinet/in.h>
+#include <sys/socket.h>
 #include <netdb.h>
 #else
 #include <Windows.h>

--- a/tensorflow/core/platform/posix/env.cc
+++ b/tensorflow/core/platform/posix/env.cc
@@ -25,6 +25,10 @@ limitations under the License.
 #include <time.h>
 #include <unistd.h>
 
+#ifdef __FreeBSD__
+#include <pthread_np.h>
+#endif
+
 #include <thread>
 #include <vector>
 
@@ -110,10 +114,7 @@ class PosixEnv : public Env {
     pthread_threadid_np(nullptr, &tid64);
     return static_cast<int32>(tid64);
 #elif defined(__FreeBSD__)
-    // Has to be casted to long first, else this error appears:
-    // static_cast from 'pthread_t' (aka 'pthread *') to 'int32' (aka 'int')
-    // is not allowed
-    return static_cast<int32>(static_cast<int64>(pthread_self()));
+    return pthread_getthreadid_np();
 #else
     return static_cast<int32>(pthread_self());
 #endif
@@ -133,7 +134,12 @@ class PosixEnv : public Env {
     return false;
 #else
     char buf[100];
+#ifdef __FreeBSD__
+    int res = 0;
+    pthread_get_name_np(pthread_self(), buf, static_cast<size_t>(100));
+#else
     int res = pthread_getname_np(pthread_self(), buf, static_cast<size_t>(100));
+#endif
     if (res != 0) {
       return false;
     }

--- a/tensorflow/lite/kernels/internal/BUILD
+++ b/tensorflow/lite/kernels/internal/BUILD
@@ -605,6 +605,9 @@ cc_library(
         ":darwin_x86_64": [
             ":neon_tensor_utils",
         ],
+        ":freebsd": [
+            ":neon_tensor_utils",
+        ],
         "//conditions:default": [
             ":portable_tensor_utils",
         ],

